pretyping/inferCumulativity.ml

```diff
@@ -0,0 +1,286 @@
+(************************************************************************)
+(*  v      *   The Coq Proof Assistant  /  The Coq Development Team     *)
+(* <O___,, *   INRIA - CNRS - LIX - LRI - PPS - Copyright 1999-2017     *)
+(*   \VV/  **************************************************************)
+(*    //   *      This file is distributed under the terms of the       *)
+(*         *       GNU Lesser General Public License Version 2.1        *)
+(************************************************************************)
+
+open Reduction
+open Declarations
+open Constr
+open Univ
+open Util
+
+let rec break_app c =
+  match Constr.kind c with
+  | App (c, args) ->
+    let hd, args0 = break_app c in
+    hd, Array.append args0 args
+  | Cast (c, _, _) -> break_app c
+  | _ -> c, [||]
+
+let infer_level_eq u variances =
+  if LMap.mem u variances
+  then LMap.set u Variance.Invariant variances
+  else variances
+
+let infer_level_leq u variances =
+  match LMap.find u variances with
+  | exception Not_found -> variances
+  | varu -> LMap.set u (Variance.sup varu Variance.Covariant) variances
+
+let infer_generic_instance_eq variances u =
+  Array.fold_left (fun variances u -> infer_level_eq u variances)
+    variances (Instance.to_array u)
+
+let variance_pb cv_pb var =
+  let open Variance in
+  match cv_pb, var with
+  | _, Irrelevant -> Irrelevant
+  | _, Invariant -> Invariant
+  | CONV, Covariant -> Invariant
+  | CUMUL, Covariant -> Covariant
+
+let infer_cumulative_ind_instance cv_pb cumi variances u =
+  Array.fold_left2 (fun variances varu u ->
+      match LMap.find u variances with
+      | exception Not_found -> variances
```

Comment:
What doesn’t this mean?

Reply:
What does this mean? (Bad completion)

Reply:
Universes not in the map are global universes.
```coq
Inductive Box2@{+u +v} := box2 : Type@{u} -> Type@{v} -> Box2.
Universe u.
Inductive Box2'@{+v} := box2' : Box2@{u (* <- this one *) v} -> Box2'.
```

Reply:
BTW, note that the match exception construct is only available from OCaml 4.03 onwards. I don't think we decided to bump the OCaml version for Coq 8.8 yet, so your patch probably needs to be tweaked (or alternatively the version constraint).

Reply:
>BTW, note that the match exception construct is only available from OCaml 4.03 onwards. I don't think we decided to bump the OCaml version for Coq 8.8 yet, so your patch probably needs to be tweaked (or alternatively the version constraint).

But CI is passing with 4.02.3?

Reply:
Oh yes, you're right. This was introduced in 4.02, my bad. Which means I strongly refrained from using this nice syntax all this time for nothing. Darn!